KB POC (#1320)
temporalio/documentation: kb/temporal-platform-limits-sheet.md

---
slug: temporal-platform-limits-sheet
title: Temporal Platform limits sheet
tags:
  - kb-article
date: 2022-06-07T00:00:00Z
---

Running into limits can cause unexpected failures, so be mindful when you design your systems.

> Here is a list of many hard (error) or soft (warn) limits that you could run into while using the Temporal Platform.

Reviewer: ```suggestion
Here is a list of many hard (error) or soft (warn) limits that you could encounter while using the Temporal Platform.
```

<!-- truncate -->

> - **gRPC**: gRPC has a limit of 4 MB for ([each message received](https://github.com/grpc/grpc/blob/v1.36.2/include/grpc/impl/codegen/grpc_types.h#L466)).

Reviewer: ```suggestion
- **gRPC**: gRPC has a limit of 4 MB for [each message received](https://github.com/grpc/grpc/blob/v1.36.2/include/grpc/impl/codegen/grpc_types.h#L466).
```

> - **Event Batch Size**: The `DefaultTransactionSizeLimit` limit is [4 MB](https://github.com/temporalio/temporal/pull/1363).

Reviewer: ```suggestion
- **Event batch size**: The `DefaultTransactionSizeLimit` limit is [4 MB](https://github.com/temporalio/temporal/pull/1363).
```

> - **gRPC**: gRPC has 4 MB size limit ([per each message received](https://github.com/grpc/grpc/blob/v1.36.2/include/grpc/impl/codegen/grpc_types.h#L466))
> - **Event Batch Size**: The `DefaultTransactionSizeLimit` limit is [4 MB](https://github.com/temporalio/temporal/pull/1363).
>   This is the largest transaction size we allow for event histories to be persisted.

Reviewer: ```suggestion
- **gRPC**: gRPC has a limit of 4 MB for ([each message received](https://github.com/grpc/grpc/blob/v1.36.2/include/grpc/impl/codegen/grpc_types.h#L466)).
- **Event Batch Size**: The `DefaultTransactionSizeLimit` limit is [4 MB](https://github.com/temporalio/temporal/pull/1363).
  This is the largest transaction size we allow for Event Histories to be persisted.
```

  - This is configurable with `TransactionSizeLimit`, if you know what you are doing.

> - **Blob size limit**: For incoming payloads (including Workflow context) - _[source](https://github.com/temporalio/temporal/blob/v1.7.0/service/frontend/service.go#L133-L134)_

Reviewer: ```suggestion
- **Blob size limit** for incoming payloads (including Workflow context; _[source](https://github.com/temporalio/temporal/blob/v1.7.0/service/frontend/service.go#L133-L134)_):
```

>   - we warn at 512 KB: [`Blob size exceeds limit.`](https://github.com/temporalio/temporal/blob/fee1c43823699e90b330680a8efeb9d8dbee8cf3/common/util.go#L568)
>   - we error at 2 MB: `ErrBlobSizeExceedsLimit: Blob data size exceeds limit.`

Reviewer: ```suggestion
  - We warn at 512 KB: [`Blob size exceeds limit.`](https://github.com/temporalio/temporal/blob/fee1c43823699e90b330680a8efeb9d8dbee8cf3/common/util.go#L568)
  - We error at 2 MB: `ErrBlobSizeExceedsLimit: Blob data size exceeds limit.`
```

  - This is configurable with [`BlobSizeLimitError` and `BlobSizeLimitWarn`](https://github.com/temporalio/temporal/blob/v1.7.0/service/history/configs/config.go#L378-L379), if you know what you are doing.

> - **History total size limit**: (leading to a terminated Workflow Execution)

Reviewer: ```suggestion
- **History total size limit** (leading to a terminated Workflow Execution):
```

  - We warn at 10 MB: `history size exceeds warn limit.`
  - We error at 50 MB: [`history size exceeds error limit.`](https://github.com/temporalio/temporal/blob/v1.7.0/service/history/workflowExecutionContext.go#L1204)
  - This is configurable with [`HistorySizeLimitError` and `HistorySizeLimitWarn`](https://github.com/temporalio/temporal/blob/v1.7.0/service/history/configs/config.go#L380-L381), if you know what you are doing.

> - **History total count limit**: (leading to a terminated Workflow Execution)
>   - We warn at 10,000 events: `history size exceeds warn limit.`
>   - We error at 50,000 events: [`history size exceeds error limit.`](https://github.com/temporalio/temporal/blob/v1.7.0/service/history/workflowExecutionContext.go#L1204)

Reviewer: ```suggestion
- **History total count limit** (leading to a terminated Workflow Execution):
  - We warn at 10,000 Events: `history size exceeds warn limit.`
  - We error at 50,000 Events: [`history size exceeds error limit.`](https://github.com/temporalio/temporal/blob/v1.7.0/service/history/workflowExecutionContext.go#L1204)
```
